### ISSUE TITLE
Teku slow to shutdown

### ISSUE THREAD

**Issue by benjaminion** [Consensys-Incorporated/teku#3341]
A couple of reports of Teku being very slow to shutdown on `^C`.

Here's one. There's reportedly a long pause (~ a minute) before `Thread pool did not terminate`. This is running in our Docker image.
```
16:51:48.438 INFO  - Sync Event  *** Current slot: 1457, Head slot: 38, Connected peers: 0
^CTeku is shutting down
16:51:51.272 FATAL - Exiting due to fatal error in Eth1DepositManager
java.util.concurrent.CompletionException: tech.pegasys.teku.service.serviceutils.FatalServiceFailureException: java.util.concurrent.CompletionException: java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.CompletableFuture$AsyncRun@7f6f4822 rejected from java.util.concurrent.ThreadPoolExecutor@295b4f83[Shutting down, pool size = 1, active threads = 1, queued tasks = 0, completed tasks = 118]
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:990) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:970) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:152) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883) ~[?:?]
        at java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2315) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:503) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:28) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateResult(SafeFuture.java:147) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$exceptionallyCompose$25(SafeFuture.java:347) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:152) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137) ~[?:?]
        at org.web3j.utils.Async.lambda$run$1(Async.java:38) ~[core-4.6.2.jar:?]
        at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1800) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) [?:?]
        at java.lang.Thread.run(Thread.java:832) [?:?]
Caused by: tech.pegasys.teku.service.serviceutils.FatalServiceFailureException: java.util.concurrent.CompletionException: java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.CompletableFuture$AsyncRun@7f6f4822 rejected from java.util.concurrent.ThreadPoolExecutor@295b4f83[Shutting down, pool size = 1, active threads = 1, queued tasks = 0, completed tasks = 118]
        at tech.pegasys.teku.pow.Eth1DepositManager.lambda$start$3(Eth1DepositManager.java:74) ~[teku-pow-20.11.1.jar:20.11.1]
        at java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:986) ~[?:?]
        ... 25 more
Caused by: java.util.concurrent.CompletionException: java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.CompletableFuture$AsyncRun@7f6f4822 rejected from java.util.concurrent.ThreadPoolExecutor@295b4f83[Shutting down, pool size = 1, active threads = 1, queued tasks = 0, completed tasks = 118]
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1155) ~[?:?]
        ... 24 more
Caused by: java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.CompletableFuture$AsyncRun@7f6f4822 rejected from java.util.concurrent.ThreadPoolExecutor@295b4f83[Shutting down, pool size = 1, active threads = 1, queued tasks = 0, completed tasks = 118]
        at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2057) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:827) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1357) ~[?:?]
        at java.util.concurrent.CompletableFuture.asyncRunStage(CompletableFuture.java:1814) ~[?:?]
        at java.util.concurrent.CompletableFuture.runAsync(CompletableFuture.java:2023) ~[?:?]
        at org.web3j.utils.Async.run(Async.java:33) ~[core-4.6.2.jar:?]
        at org.web3j.protocol.Service.sendAsync(Service.java:60) ~[core-4.6.2.jar:?]
        at org.web3j.protocol.core.Request.sendAsync(Request.java:91) ~[core-4.6.2.jar:?]
        at tech.pegasys.teku.pow.contract.DepositContract.depositEventInRange(DepositContract.java:139) ~[teku-pow-20.11.1.jar:20.11.1]
        at tech.pegasys.teku.pow.DepositFetcher.processDepositsInBatch(DepositFetcher.java:128) ~[teku-pow-20.11.1.jar:20.11.1]
        at tech.pegasys.teku.pow.DepositFetcher.sendNextBatchRequest(DepositFetcher.java:92) ~[teku-pow-20.11.1.jar:20.11.1]
        at tech.pegasys.teku.pow.DepositFetcher.lambda$sendNextBatchRequest$2(DepositFetcher.java:120) ~[teku-pow-20.11.1.jar:20.11.1]
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1146) ~[?:?]
        ... 24 more
16:51:51.323 WARN  - Error shutting down connection manager
16:51:51.326 INFO  - Stopping discovery server
16:51:51.327 INFO  - Shutting down discovery server
16:51:53.334 INFO  - Scheduler DefaultQuartzScheduler_$_NON_CLUSTERED shutting down.
16:51:53.334 INFO  - Scheduler DefaultQuartzScheduler_$_NON_CLUSTERED paused.
16:51:53.335 INFO  - Scheduler DefaultQuartzScheduler_$_NON_CLUSTERED shutdown complete.
Thread pool did not terminate
```

**Comment by ajsutton** [comment]
Teku is waiting on a read from the eth1 node which is preventing shutdown:

```
"pool-18-thread-1" #88 prio=5 os_prio=31 cpu=136.08ms elapsed=280.66s tid=0x00007f9912b2f000 nid=0x15d03 runnable  [0x0000700013863000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.Net.poll(java.base@14.0.2/Native Method)
	at sun.nio.ch.NioSocketImpl.park(java.base@14.0.2/NioSocketImpl.java:181)
	at sun.nio.ch.NioSocketImpl.timedRead(java.base@14.0.2/NioSocketImpl.java:285)
	at sun.nio.ch.NioSocketImpl.implRead(java.base@14.0.2/NioSocketImpl.java:309)
	at sun.nio.ch.NioSocketImpl.read(java.base@14.0.2/NioSocketImpl.java:350)
	at sun.nio.ch.NioSocketImpl$1.read(java.base@14.0.2/NioSocketImpl.java:803)
	at java.net.Socket$SocketInputStream.read(java.base@14.0.2/Socket.java:982)
	at okio.InputStreamSource.read(JvmOkio.kt:90)
	at okio.AsyncTimeout$source$1.read(AsyncTimeout.kt:129)
	at okio.RealBufferedSource.indexOf(RealBufferedSource.kt:449)
	at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.kt:333)
	at okhttp3.internal.http1.HeadersReader.readLine(HeadersReader.kt:29)
	at okhttp3.internal.http1.Http1ExchangeCodec.readResponseHeaders(Http1ExchangeCodec.kt:178)
	at okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.kt:106)
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.kt:79)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:34)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
	at okhttp3.internal.connection.RealCall.execute(RealCall.kt:154)
	at org.web3j.protocol.http.HttpService.performIO(HttpService.java:160)
	at org.web3j.protocol.Service.send(Service.java:48)
	at org.web3j.protocol.Service.lambda$sendAsync$0(Service.java:60)
	at org.web3j.protocol.Service$$Lambda$617/0x00000008010a1840.call(Unknown Source)
	at org.web3j.utils.Async.lambda$run$1(Async.java:38)
	at org.web3j.utils.Async$$Lambda$618/0x00000008010a1c40.run(Unknown Source)
	at java.util.concurrent.CompletableFuture$AsyncRun.run(java.base@14.0.2/CompletableFuture.java:1800)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@14.0.2/ThreadPoolExecutor.java:1130)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@14.0.2/ThreadPoolExecutor.java:630)
	at java.lang.Thread.run(java.base@14.0.2/Thread.java:832)

```

**Comment by rolfyone** [comment]
Geth is taking a while for the socket to time out and close, and despite efforts being made it's not easy to close from our side. We might just have to live with this for now, as there doesn't seem to be a good work-around.

I guess, closing as "won't fix" unless more information comes to light to suggest it's more than just a slow shutdown.
